### PR TITLE
+add collectType by @hepin1989

### PR DESCRIPTION
**Pull request by ktoso**
Push https://github.com/akka/akka/pull/24258 over the finish line, see my commit for additions.

Added docs, tests, and yes I did replicate the method to scaladsl -- for both consistency between APIs and personal often "i want this" during presentations I did. Hope that's in line with the theme that we want to add convenience, and here it is done at a very low cost I believe.

### CODE REVIEW

**Comment by ktoso** [comment]
Resolves refs #24254

**Comment by ktoso** [comment]
Please merge as rebase, so authorship is kept for @hepin1989

File: akka-stream/src/main/mima-filters/2.5.11.backwards.excludes [hunk 1]
```diff
@@ -0,0 +1 @@
+ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.stream.scaladsl.FlowOps.collectType")
```
Comment (patriknw) [inline akka-stream/src/main/mima-filters/2.5.11.backwards.excludes]: that file name is off by 2, should be 2.5.9
https://github.com/akka/akka/blob/master/CONTRIBUTING.md#binary-compatibility

File: project/AutomaticModuleName.scala [hunk 2]
```diff
@@ -17,7 +18,10 @@ import sbt.Keys._
...
   private val AutomaticModuleName = "Automatic-Module-Name"
 
-  def settings(name: String): Seq[Def.Setting[Task[Seq[PackageOption]]]] = Seq(
-    packageOptions in (Compile, packageBin) += Package.ManifestAttributes(AutomaticModuleName → name)
+  def settings(name: String): Seq[Def.Setting[_]] = Seq(
+    packageOptions in (Compile, packageBin) += Package.ManifestAttributes(AutomaticModuleName → name),
+    // explicitly pass it to the osgi plugin, otherwise it would lose that setting
+    // This should be likely fixed in sbt-osgi, such that it does pick up package options as additional headers
+    OsgiKeys.additionalHeaders := Map(AutomaticModuleName → name)
   )
```
Comment (patriknw) [inline project/AutomaticModuleName.scala]: why is this included in this PR?

**Review by patriknw — CHANGES_REQUESTED** [review]
looking good, but...

File: project/AutomaticModuleName.scala (hunk 2 above)
Comment (ktoso) [reply project/AutomaticModuleName.scala]: It should not be, seems my master branch included it... fixing

**Comment by ktoso** [comment]
Fixed things and squashed

File: akka-stream-tests/src/test/java/akka/stream/javadsl/FlowTest.java [hunk 3]
```diff
@@ -10,12 +10,14 @@
...
 import akka.japi.function.*;
 import akka.stream.*;
+import akka.stream.scaladsl.FlowSpec;
 import akka.util.ConstantFun;
 import akka.stream.javadsl.GraphDSL.Builder;
 import akka.stream.stage.*;
 import akka.testkit.AkkaSpec;
 import akka.stream.testkit.TestPublisher;
 import akka.testkit.javadsl.TestKit;
+import com.google.common.collect.Lists;
```
Comment (patriknw) [inline akka-stream-tests/src/test/java/akka/stream/javadsl/FlowTest.java]: wrong import, and not used

File: akka-stream/src/main/scala/akka/stream/javadsl/SubSource.scala [hunk 4]
```diff
@@ -294,6 +294,27 @@ class SubSource[+Out, +Mat](delegate: scaladsl.SubFlow[Out, Mat, scaladsl.Source
...
+   *
+   * '''Completes when''' upstream completes
+   *
+   * '''Cancels when''' downstream cancels
+   */
+  def collectType[T](clazz: Class[T]): javadsl.SubSource[T, Mat] = {
+    new SubSource(delegate.collect({
+      case elem if clazz.isInstance(elem) ⇒ clazz.cast(elem)
+    }))
+  }
```
Comment (2m) [inline akka-stream/src/main/scala/akka/stream/javadsl/SubSource.scala]: Why not call `collectType` on the `delegate`?

File: akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala [hunk 5]
```diff
@@ -591,6 +591,27 @@ final class Flow[-In, +Out, +Mat](delegate: scaladsl.Flow[In, Out, Mat]) extends
...
+   *
+   * '''Completes when''' upstream completes
+   *
+   * '''Cancels when''' downstream cancels
+   */
+  def collectType[T](clazz: Class[T]): javadsl.Flow[In, T, Mat] = {
+    new Flow(delegate.collect({
+      case elem if clazz.isInstance(elem) ⇒ clazz.cast(elem)
+    }))
+  }
```
Comment (2m) [inline akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala]: Here as well delegate to `delegate`.

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowSpec.scala [hunk 6]
```diff
@@ -25,7 +28,11 @@ import akka.stream.impl.fusing.GraphInterpreterShell
 object FlowSpec {
   class Fruit
   class Apple extends Fruit
-  val apples = () ⇒ Iterator.continually(new Apple)
+  class Orange extends Fruit
+  val fruits = () ⇒ new Iterator[Fruit] {
+    override def hasNext: Boolean = true
+    override def next(): Fruit = if (ThreadLocalRandom.current().nextBoolean()) new Apple else new Orange
+  }
```
Comment (2m) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowSpec.scala]: Was the intention of this modification so that `FlowSpec.fruits` could be used on `collectType` tests?

**Review by 2m — APPROVED** [review]
LGTM. Just a couple of nitpicks.

**Review by patriknw — APPROVED** [review]
LGTM

File: akka-stream/src/main/scala/akka/stream/javadsl/SubSource.scala (hunk 4 above)
Comment (ktoso) [reply akka-stream/src/main/scala/akka/stream/javadsl/SubSource.scala]: thanks done

File: akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala (hunk 5 above)
Comment (ktoso) [reply akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala]: done

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowSpec.scala (hunk 6 above)
Comment (ktoso) [reply akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowSpec.scala]: Mostly wanted the fruits collection in:

```
"be covariant" in {
      val f1: Source[Fruit, _] = Source.fromIterator[Fruit](fruits)
      val p1: Publisher[Fruit] = Source.fromIterator[Fruit](fruits).runWith(Sink.asPublisher(false))
```

be a real fruits collection

**Comment by ktoso** [comment]
Ehhh... formatting of braces...

**Review by johanandren — APPROVED** [review]

**Comment by ktoso** [comment]
Failures unrelated https://github.com/akka/akka/issues/24579 and https://github.com/akka/akka/issues/24580

**Comment by ktoso** [comment]
Thanks @hepin1989 !

**Comment by hepin1989** [comment]
Sorry，it's my shame~~~, Moving around during the Spring Festival.
